File: sorting-in-linear-time/python/counting_sort.py
import random


def sort(data):
    counting_table = [0 for _ in range(max(data) + 1)]

    for d in data:
        counting_table[d] += 1
    for i in range(len(counting_table) - 1):
        counting_table[i + 1] += counting_table[i]

    print("--- counting table ---")
    print(counting_table)

    temp = [-1 for _ in data]
    for i in reversed(range(len(data))):
        idx_to_replace = counting_table[data[i]] - 1
        temp[idx_to_replace] = data[i]
        counting_table[data[i]] -= 1

    data.clear()
    data.extend(temp)


if __name__ == "__main__":
    data = [random.randint(1, 20) for _ in range(40)]

    print("--- before ---")
    print(data)

    print("--- sort ---")
    sort(data)

    print("--- after ---")
    print(data)

File: sorting-in-linear-time/golang/counting_sort.go
package main

import (
	"fmt"
	"math/rand"
	"slices"
)

func sort(data []int) {
	countingTable := make([]int, slices.Max(data)+1)

	for _, d := range data {
		countingTable[d] += 1
	}
	for i := 0; i < len(countingTable)-1; i++ {
		countingTable[i+1] += countingTable[i]
	}

	fmt.Println("--- counting table ---")
	fmt.Println(countingTable)

	temp := make([]int, len(data))
	for i := len(data) - 1; i >= 0; i-- {
		idxToReplace := countingTable[data[i]] - 1
		temp[idxToReplace] = data[i]
		countingTable[data[i]] -= 1
	}

	clear(data)
	data = data[:0]

	data = append(data, temp...)
}

func main() {
	data := make([]int, 40)

	for i := range 40 {
		data[i] = rand.Intn(20) + 1
	}

	fmt.Println("--- before ---")
	fmt.Println(data)

	fmt.Println("--- sort ---")
	sort(data)

	fmt.Println("--- after ---")
	fmt.Println(data)
}
